src/random/wyrand.cr

```diff
@@ -0,0 +1,63 @@
+require "random/secure"
+
+# A native Crystal implementation of the beautifully simple and impressively
+# fast PRNG from Wyhash.
+#
+# Based on the original C implementation by 王一 (Wang Yi).
+# https://github.com/wangyi-fudan/wyhash
+#
+# Provides a 64 bit output based on mix and multiply operation. This provides
+# exceptionly fast performance on modern processors, but may yield slower
+# results on older architectures.
+struct Random::Wyrand
+  include Random
+
+  private P = {0xa0761d6478bd642f_u64, 0xe7037ed1a0b428db_u64}
+
+  @state : UInt64
+
+  def initialize(seed = Random::Secure.rand(UInt64))
+    @state = seed
+  end
+
+  def next_u : UInt64
+    a = @state &+= P[0]
+    b = a ^ P[1]
+
+    r = UInt128.new! a
+    r &*= b
+
+    p = pointerof(r).as UInt64*
+    a = p.value
+    b = (p + 1).value
+
+    a ^ b
+  end
+
+  {% for size in [8, 16, 32, 64] %}
+    {% type = "UInt#{size}".id %}
+
+    def rand(type : {{type}}.class) : {{type}}
+      r = next_u
+      p = pointerof(r).as {{type}}*
+      p.value
+    end
+
+    private def rand_int(max : {{type}}) : {{type}}
+      a = next_u
+      b = UInt64.new! max
+
+      r = UInt128.new! a
+      r &*= b
+
+      p = pointerof(r).as {{type}}*
+      (p + {{128 // size - 1}}).value
```

Comment:
why this pointer arithmetic matters? Isn't right shift just enough?

Reply:
Ahh, I understand. But then calculation is wrong.
Fix me if I'm mistaken, but seems like it will be simpler:
```crystal
  r &*= b
  {{type}}.new!(r >> 64)
```

Reply:
This is most definitely an error. The correct implementation using pointer arithmetic is
```crystal
(p + {{128 // size // 2}}).value
```
The specs (both here) and importantly in std lib do not look to check distribution on these outputs. I expect this would be significantly biased towards 0.